### ISSUE TITLE
[REVIEW]: PyBioPAX: biological pathway exchange in Python

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#4136]
**Submitting author:** @bgyori (<a href="http://orcid.org/0000-0001-9439-5346">Benjamin M. Gyori</a>)
**Repository:** https://github.com/indralab/pybiopax
**Branch with paper.md** (empty if default branch): 
**Version:** 0.0.1
**Editor:** @majensen
**Reviewers:** @dhimmel, @leonweber
**Archive:** 10.5281/zenodo.6331584

 

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/25f800a45d8d0683e73f9e0e75883dee/status.svg)](https://joss.theoj.org/papers/25f800a45d8d0683e73f9e0e75883dee)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/25f800a45d8d0683e73f9e0e75883dee"><img src="https://joss.theoj.org/papers/25f800a45d8d0683e73f9e0e75883dee/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/25f800a45d8d0683e73f9e0e75883dee/status.svg)](https://joss.theoj.org/papers/25f800a45d8d0683e73f9e0e75883dee)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/indralab/pybiopax" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@dhimmel & @leonweber, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @majensen know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

## Review checklist for @dhimmel
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/indralab/pybiopax">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/indralab/pybiopax">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@bgyori) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @leonweber
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/indralab/pybiopax">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/indralab/pybiopax">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@bgyori) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Wordcount for `paper.md` is 1074

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.15252/msb.20177651 is OK
- 10.1371/journal.pcbi.1003194 is OK
- 10.1038/nbt.1666 is OK
- 10.1093/nar/gkz946 is OK
- 10.18653/v1/2021.emnlp-main.93 is OK
- 10.1186/s12859-020-03568-5 is OK
- 10.1038/ncomms12846 is OK
- 10.1038/msb.2013.1 is OK
- 10.1371/journal.pcbi.1006220 is OK
- 10.1093/bioinformatics/btx660 is OK
- 10.1186/gb-2010-11-1-r3 is OK
- 10.1093/bib/bbx085 is OK
- 10.1093/nar/gkz1031 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.07 s (543.7 files/s, 113792.1 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                          16            489            431           1395
XML                              1             62              0            411
Jupyter Notebook                 3              0           3775            397
Markdown                         4             65              0            342
TeX                              1             14              0            170
DOS Batch                        1              8              1             26
INI                              1              4              0             26
YAML                             1              2              0             18
reStructuredText                 6             15             41             14
make                             1              4              7              9
TOML                             1              1              0              8
Bourne Shell                     1              2              1              6
-------------------------------------------------------------------------------
SUM:                            37            666           4256           2822
-------------------------------------------------------------------------------


Statistical information for the repository 'cc16f95d07fd1d3ecae464d2' was
gathered on 2022/02/07.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Ben Gyori                       91          2318            533           64.74
Charles Tapley Hoyt             47          1033            520           35.26

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Benjamin M. Gyori          1845          100.0          9.2                9.97
Charles Tapley Hoyt         470           45.5          2.1                5.96
```

**Comment by dhimmel** [comment]
# Daniel Himmelstein's Review

I completed my checklist above. Some general comments: I think it's helpful to have a Python library to interact with BioPax models. The loaders are versatile and convenient. I explored the package briefly in [this notebook](https://colab.research.google.com/drive/1iNbqZPBy02riZCIgshY_Y_sVdzHJRSZN?usp=sharing). I opened https://github.com/indralab/pybiopax/issues/28 about whether visualization is possible.

The [testing coverage](https://github.com/indralab/pybiopax/runs/4571578639?check_suite_focus=true#step:5:43) is impressive. Installation was easy via pip. Dependencies [are light](https://github.com/indralab/pybiopax/blob/aca21e51769f07a121816004dc143e048b9b9764/setup.py#L39): lxml, requests, tqdm.

I'm slightly worried about link rot for the notebook links in the manuscript, since the link to the master branch rather than using a tag or commit to pin a version:

```
https://nbviewer.org/github/indralab/pybiopax/blob/master/notebooks/Explore%20Pathway%20Commons.ipynb
https://nbviewer.org/github/indralab/pybiopax/blob/master/notebooks/Pathway%20Anticorrelation%20Analysis.ipynb
```

Overall a very valuable software contribution to make working with the cryptic BioPAX format a bit easier and more accessible.

**Comment by bgyori** [comment]
Thank you @dhimmel! I updated the links in the manuscript to use the `0.1.0` tag which should guarantee that they will be stable.

**Comment by whedon** [comment]
:wave: @leonweber, please update us on how your review is going (this is an automated reminder).

**Comment by leonweber** [comment]
# Leon Weber's Review

PyBioPAX is a python library to parse and interact with BioPAX files from python. This package closes an important gap in the python bioinformatics ecosystem, as BioPAX is a widespread standard for distributing pathway models and before PyBioPAX parsing such files in python was cumbersome. 

I have used the package [in the past](https://github.com/leonweber/extend), there a data loaders for many use cases which are very convenient. The package is extensively tested and installation is easy with few dependencies.

My only concern is that the `State of the field` section seems to be missing from the paper. I have created an issue for this (https://github.com/indralab/pybiopax/issues/30).

**Comment by bgyori** [comment]
@whedon generate pdf

**Comment by bgyori** [comment]
@editorialbot generate pdf

**Comment by bgyori** [comment]
Thank you @leonweber, we created a new dedicated subsection called State of the field under the Statement of need section where we discuss other commonly used packages for BioPAX. Let us know if anything further is needed. Thank you!

**Comment by leonweber** [comment]
Thanks for the quick response @bgyori. For me, there are no further open issues. All boxes are checked in my checklist above.

**Comment by dhimmel** [comment]
Approved on my end as well.

**Comment by bgyori** [comment]
@majensen looks like both reviewers have finished their review. What are the next steps? Thanks!

**Comment by majensen** [comment]
Thanks @bgyori - and @dhimmel, @leonweber for your reviews. I will make a editorial pass over the paper which may result in some minor comments for author consideration. Once that is done and pulled in, I'll ask for a Zenodo (or similar) archive of the repo. Stand by for those details until we have the final version. Appreciated!

**Comment by majensen** [comment]
@editorialbot generate pdf

**Comment by majensen** [comment]
@editorialbot check references

**Comment by majensen** [comment]
@bgyori this is a nice tool and appears to be a very complete representation of the underlying spec. The notebooks are detailed and give a user a way in to the practical use of this framework.

Can you please fix the following issues in the references (line numbers from the latest PDF generation above):

l. 107 - capitalize "-Parser" - per original reference
l. 147 - capitalize "Python" - ''
l. 150 - capitalize "R" - ''
l. 152 - author list is borked
l. 156 - capitalize according to original reference

l. 116 and l. 155 - handling of long author list appears not to have worked well here, and inconsistently from elsewhere (e.g., l. 131)

thx

**Comment by bgyori** [comment]
@editorialbot generate pdf

**Comment by bgyori** [comment]
@majensen thank you for your feedback! I fixed all of the issues you pointed out above.

**Comment by majensen** [comment]
Great - thanks @bgyori  - the last item before we ask the associate editors-in-chief to review is to create an open archive of the the latest state of the code repo. This can be done with Zenodo, FigShare, or similar. (Here is a brief tutorial using Zenodo: https://github.com/openjournals/joss-reviews/issues/1839#issuecomment-560048729.)
JOSS prefers that the name of the archive match that of the paper, and that the author list be the same as well. 
Once this is accomplished, could you pls put the DOI of the archive here as a comment. I will then make some settings and invoke the AEIC. 
Thanks!

**Comment by bgyori** [comment]
Hi @majensen, here is the DOI for the Zenodo archive: 10.5281/zenodo.6331584 (https://doi.org/10.5281/zenodo.6331584). Thank you!

**Comment by majensen** [comment]
@editorialbot set 0.0.1 as version

**Comment by majensen** [comment]
@editorialbot set 10.5281/zenodo.6331584 as archive

**Comment by majensen** [comment]
@editorialbot recommend-accept

**Comment by majensen** [comment]
Help @xuanxu - there was a fail in the crossref xml validation (see  https://github.com/openjournals/joss-papers/runs/5440427063?check_suite_focus=true#step:6:53). But here is the crossref element:
`<isbn>1367-4811 (Electronic)</isbn>`
so maybe the validation is too strict?

thanks

**Comment by xuanxu** [comment]
@majensen in this case it looks like that entry is incorrect. The ISBN [is invalid](https://isbnsearch.org/search?s=1367-4811) and the value should not include the `(Electronic)` part.
In the bib file the isbn and issn [are the same](https://github.com/indralab/pybiopax/blob/master/paper/paper.bib#L123-L124), I guess that is an error.

And (not related) in that same reference the [url value](https://github.com/indralab/pybiopax/blob/master/paper/paper.bib#L131) looks wrong too.

**Comment by majensen** [comment]
Thanks @xuanxu  - since the Crossref XML is created by the bot, not sure what the options would be.

**Comment by bgyori** [comment]
@majensen @xuanxu done, I fixed the problematic reference.

**Comment by majensen** [comment]
@editorialbot recommend-accept

**Comment by bgyori** [comment]
@majensen sorry to bug you, it seems like we are very close here. The artifacts seem to have generated well per the message above.

**Comment by majensen** [comment]
@bgyori. yes, I'm afraid this is in the hands of other volunteer, unpaid people with day jobs now, the @openjournals/joss-eics.

**Comment by kyleniemeyer** [comment]
@editorialbot accept

**Comment by kyleniemeyer** [comment]
Congratulations @bgyori on your article's publication in JOSS!

Many thanks to @dhimmel and @leonweber for reviewing this, and @majensen for editing.
